virtcontainers/container.go

```diff
@@ -646,6 +646,14 @@ func filterDevices(c *Container, devices []ContainerDevice) (ret []ContainerDevi
 			continue
 		}
 
+		devType := c.sandbox.devManager.GetDeviceByID(dev.ID).DeviceType()
+		if devType == config.DeviceVFIO {
```

Comment:
deviceManager.createDevice() is supposed to avoid adding the same device twice. Any reason that is not working? /cc @WeiZhang555 

Also we should just filter already added devices rather than skipping VFIO devices all together.

Reply:
Thanks,  @bergwolf.  I have read the deviceManager.createDevice.  The function dm.findDeviceByMajorMinor can avoid adding the same device. Maybe if the devices use dm.NewDevice to create, it can be working. Now the vfio device is been added to Qemu command line directly.  Please correct me if I'm wrong, I will check again, Thanks.

Yes, I will optimize the filter vfio devices :-)

Reply:
@zhabinecho whatever you filter in filterDevices() must have gone through c.sandbox.devManager.NewDevice()->dm.createDevice(). So I don't see why you end up plugging the same vfio device obviously. One possibility is that the persistent data is inconsistent. Could you check `/var/run/vc/sbs/<sandbox-id>/devices.json` and see if the failed vfio device is listed there? You can add a long sleep where the qmp command fails and then check the devices.json file.

Reply:
> Now the vfio device is been added to Qemu command line directly.

@zhabinecho All vfio devices should be hotplugged and not passed on the command line. Are you seeing otherwise?
As @bergwolf pointed out, this likely seems an issue with persisting data.

Reply:
@zhabinecho Is this a network device? And are you using an sriov network plugin?

Reply:
@amshinde Yes, It is network device which is physical endpoint created by CNI network plugin. @bergwolf is right, the vfio device hasn't been persisted. I haven't found the right way to persist it in the physical endpoint attach.